File: Python/String/16_word_break.py
"""
Idea:
using recursion to find all possible combinations of words & searching if all exist in dictionary.
"""


def wordBreak(line, dictionary):
    if len(line) == 0:  # Base case: an empty string can always be broken.
        return 1

    for i in range(1, len(line) + 1):
        # trying all possible combination of words
        prefix = line[:i]
        # If the prefix is in the dictionary and the remaining part can be broken
        if prefix in dictionary and wordBreak(line[i:], dictionary):
            return 1
    return 0


if __name__ == '__main__':
    test_case = int(input())
    for _ in range(test_case):
        number_of_elements = int(input())
        dictionary = [word for word in input().strip().split()]
        line = input().strip()
        res = wordBreak(line, dictionary)
        if res:
            print(1)
        else:
            print(0)

File: C++/String/16_word_break.cpp
/*
Idea:
using recursion to find all possible combinations of words & searching if all exist in dictionary.
*/

//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


class Solution
{
public:
    int wordBreak(string word, vector<string>& dictionary) {
        if (word.empty())
            return 1;  // Base case: an empty string can always be broken.

        for (int i = 1; i <= word.size(); ++i)
        {
            //trying all possible combination of words
            string prefix = word.substr(0, i);
            // If the prefix is in the dictionary and the remaining part can be broken
            if (find(dictionary.begin(), dictionary.end(), prefix) != dictionary.end() && wordBreak(word.substr(i), dictionary))
                return 1;
        }

        return 0;
    }

};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<string> dict;
        for (int i = 0;i < n;i++) {
            string S;
            cin >> S;
            dict.push_back(S);
        }
        string line = "lrbbmqbabowkkab";
        cin >> line;
        Solution ob;
        cout << ob.wordBreak(line, dict) << "\n";
    }
}
